### DISCUSSION TITLE
Hidden LeaseSet for SSH tunnel

### DISCUSSION THREAD

**Discussion by spencermp** [PurpleI2P/i2pd#2104] [Q&A] [answered]
Are there any docs on using encrypted lease sets for a private SSH tunnel with i2pd? 
There are some nice tutorials for the Java router but the documentation seems lacking on the i2pd front.

I currently have a configuration that looks like this:
```
# server config on host computer
[ssh-serv]
type = server
host = 127.0.0.1
port = 22
inbound.length = 1
outbound.length = 1
signaturetype = 11
i2cp.leaseSetType = 5
keys = ssh-server.dat

# client config on guest computer
type = client
host = 127.0.0.1
port = 1337
inbound.length = 1
outbound.length = 1
i2cp.dontPublishLeaseSet = true
destination = <b33 address from server's i2pd web>
keys = ssh-client.dat
```

If I try to connect I get the following result. I have confirmed that direct ssh works, so this is purely an I2P thing. 
```
$ ssh -p 1337 server@127.0.0.1
kex_exchange_identification: read: Connection reset by peer
Connection reset by 127.0.0.1 port 1337
```

Any help is greatly appreciated, and thank you.

**Comment by spencermp** [comment, marked as answer]
I've figured it out. 
The one line I was missing
```
i2cp.leaseSetPrivKey = <some b64 string>
```

So the full config is now 
```
# server config on host computer
[ssh-serv]
type = server
host = 127.0.0.1
port = 22
inbound.length = 1
outbound.length = 1
signaturetype = 11
i2cp.leaseSetType = 5
# decodes to "private key example"
i2cp.leaseSetPrivKey = cHJpdmF0ZSBrZXkgZXhhbXBsZQo=
keys = ssh-server.dat

# client config on guest computer
type = client
host = 127.0.0.1
port = 1337
inbound.length = 1
outbound.length = 1
i2cp.dontPublishLeaseSet = true
i2cp.leaseSetPrivKey = cHJpdmF0ZSBrZXkgZXhhbXBsZQo=
destination = <b33 address from server's i2pd web>
keys = ssh-client.dat
```

**Reply by mashdragon** [reply]
`i2pd` says that private key is not in the right format: `critical - Destination: Invalid value i2cp.leaseSetPrivKey: cHJpdmF0ZSBrZXkgZXhhbXBsZQo=`

How do we give it a key in the right format?

**Reply by anikey-from-i2p** [reply]
Instead of encrypted leasesets, I use blinded leasesets, which works for me.

I set the following options for the tunnel:
```
signaturetype = 11
i2cp.leaseSetType = 5
i2cp.leaseSetEncType = 4
```

You can see them documented [here](https://i2pd.readthedocs.io/en/latest/user-guide/tunnels/#signature-types).

(Probably the last one is not necessary since it is default)

**Reply by mashdragon** [reply]
Thank you! We don't have to set `i2cp.dontPublishLeaseSet` to use blinded leasesets, right?

I think another user suggested this to me too: https://github.com/PurpleI2P/i2pd_docs_en/issues/101. Strangely I have chanced a bug and i2pd is crashing on the server side not long after it gets a successful connection from a client, https://github.com/PurpleI2P/i2pd/issues/2124

**Reply by anikey-from-i2p** [reply]
> We don't have to set `i2cp.dontPublishLeaseSet` to use blinded leasesets, right?

As I understand it, if you enable this option, it won't publish the leaseset at all, so it wouldn't be reachable at all. So you don't need to set it (i don't set it either). It would probably be useful for client tunnels, not server tunnels.

> I think another user suggested this to me too: [PurpleI2P/i2pd_docs_en#101](https://github.com/PurpleI2P/i2pd_docs_en/issues/101). Strangely I have chanced a bug and i2pd is crashing on the server side not long after it gets a successful connection from a client, #2124

I use blinded leasesets and have not experienced any crashes of i2pd related to their usage, neither on client side nor server side.

**Reply by mashdragon** [reply]
If you use these exact settings https://github.com/PurpleI2P/i2pd/issues/2124 it crashed for me. But orignal pointed out that dontPublishLeaseSet shouldn't be used in a server tunnel, so removing that plus removing out the speed limits got it not to crash anymore.
